datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -138,11 +167,78 @@ impl GroupedHashAggregateStreamV2 {
             aggr_layout,
             baseline_metrics,
             aggregate_expressions,
-            aggr_state: Default::default(),
+            aggr_state,
             random_state: Default::default(),
             batch_size,
             row_group_skip_position: 0,
-        })
+        };
+
+        let stream = futures::stream::unfold(inner, |mut this| async move {
+            let elapsed_compute = this.baseline_metrics.elapsed_compute();
+
+            loop {
+                let result: ArrowResult<Option<RecordBatch>> =
+                    match this.input.next().await {
+                        Some(Ok(batch)) => {
+                            let timer = elapsed_compute.timer();
+                            let result = group_aggregate_batch(
+                                &this.mode,
+                                &this.random_state,
+                                &this.group_by,
+                                &mut this.accumulators,
+                                &this.group_schema,
+                                this.aggr_layout.clone(),
+                                batch,
+                                &mut this.aggr_state,
+                                &this.aggregate_expressions,
+                            )
+                            .await;
+
+                            timer.done();
+
+                            match result {
+                                Ok(_) => continue,
```

Comment:
IMHO, this would be place to do something like: 

```rust
 Ok(_) => {
    let new_data_size = this.aggr_state.get_current_size();
    let acquired = this.memory_manager.can_grow_directly(new_data_size - data_size_before_batch, data_size_before_batch);
    if !acquired {
        this.aggr_state.spill();
        this.memory_manager.record_free_then_acquire(data_size, 0);
    }
    continue;
}
```
we basically assume  that `group_aggregate_batch` can get all the memory it needs, no need to do per row interaction with memory manager.

this would decouple process and accounting

Reply:
The interaction is not per row. It's per batch. I can place the accounting here. The code you propose is basically the same that currently runs, just inlined (it's the default impl. of `MemoryConsumer::try_grow`).

Reply:
Apologies you're right @crepererum it is per batch.

The reason why I believe moving it out makes sense is separation of concerns, but it's up to you.

for example, at line 363

```rust
        // allocate memory
        // This happens AFTER we actually used the memory, but simplifies the whole accounting and we are OK with
        // overshooting a bit. Also this means we either store the whole record batch or not.
        memory_consumer.alloc(allocated).await?;
```
can this trigger spill? will the state be consistent if spill is triggered. My guess it will be not, it might be implementation specific, but hard to tell without understanding memory management implementation, and store implementation.